nixpkgs/nixpkgs.bzl

```diff
@@ -85,40 +89,56 @@ def _expand_location(repository_ctx, string, labels, attr = None):
     Returns:
       The string with all instances of `$(location )` replaced by paths.
     """
-    num = string.count("$(location ")
     result = ""
     offset = 0
-    for i in range(num):
-        start = string.find("$(location ", offset)
-        label_start = start + len("$(location ")
-        label_end = string.find(")", label_start)
-        if label_end == -1:
-            fail("Unbalanced parentheses in location expansion for '{}'.".format(string[start:]), attr)
-        end = label_end + 1
-        label_str = string[label_start:label_end]
-        label_candidates = [
-            (lbl, path)
-            for (lbl, path) in labels.items()
-            if lbl.relative(label_str) == lbl
-        ]
-        if len(label_candidates) == 0:
-            fail("Unknown label '{}' in location expansion for '{}'.".format(label_str, string), attr)
-        elif len(label_candidates) > 1:
-            fail(
-                "Ambiguous label '{}' in location expansion for '{}'. Candidates: {}".format(
-                    label_str,
-                    string,
-                    ", ".join([str(lbl) for lbl in label_candidates]),
-                ),
-                attr,
-            )
-        location = paths.join(".", paths.relativize(
-            str(repository_ctx.path(label_candidates[0][1])),
-            str(repository_ctx.path(".")),
-        ))
-        result += string[offset:start] + location
-        offset = end
-    result += string[offset:]
+    # Step through occurrences of `$`. This is bounded by the length of the string.
+    for _ in range(len(string)):
```

Comment:
This function is a bit too unwieldy for my taste.

I would have said we should split it up into generators, but skylark does not have them. However, the location strings should be short enough for it not to matter.

So e.g. have one function which returns all “`$` to end” strings for $ that are not escaped, one function tries to parse the location syntax, and then one function which replaces the location syntax with its actual content.